mountpoint-s3/src/prefetch.rs

```diff
@@ -461,9 +461,7 @@ where
             return Ok(false);
         };
         let future_remaining = self.future_tasks.iter().map(|task| task.remaining()).sum::<usize>() as u64;
-        if total_seek_distance
-            >= (current_task.remaining() as u64 + future_remaining).min(self.config.max_forward_seek_distance)
-        {
+        if total_seek_distance >= (current_task.remaining() as u64 + future_remaining) {
```

Comment:
Might be cleaner to just make all the seeking decisions here. I think it's something like this, imagining tasks have a `available()` method that returns how many bytes are available to read right now:
```rust
let mut available = 0;
for task in std::iter::once(&*current_task).chain(self.future_tasks.iter()) {
    available += task.available();
    // If this request isn't fully downloaded then we can't seek to bytes in any later ones
    if task.available() != task.remaining() {
        break;
    }
};

let future_remaining = self.future_tasks.iter().map(|task| task.remaining()).sum::<usize>() as u64;
let remaining = current_task.remaining() + future_remaining;

// If we have enough bytes already downloaded (`available`) to skip straight to this read, then do
// it. Otherwise, we're willing to skip up to `max_forward_seek_distance` ahead without aborting the
// inflight requests.
let max_seek_distance = remaining.min(self.max_forward_seek_distance).max(available);

if total_seek_distance > max_seek_distance {
    return Ok(false);
}
```

Reply:
>  // If this request isn't fully downloaded then we can't seek to bytes in any later ones

It may be an edge case (we only have 2 requests max atm), but I don't see why introduce this limitation: if we are seeking to the range of a `future_task` request, why would it matter if the one in `current_task` isn't complete?

Reply:
I think we can rewrite this logic to only have a single loop through the tasks to find the one where the **requested** range contains the seek offset. Then, we can apply the check for the `available()` bytes with the `max_forward_seek_distance` tolerance window.

Also, I suspect it would be clearer if we reframe everything in terms of absolute offsets (or ranges), rather than `reamaining` or `available` bytes.

Reply:
Something roughly like this:

```rust
// Jump ahead to the right request
if offset >= current_task.end_offset() {
    self.next_sequential_read_offset = current_task.end_offset();
    self.current_task = None;
    while let Some(next_request) = self.future_tasks.pop_front() {
        if next_request.end_offset() > offset {
            self.current_task = Some(next_request);
            break;
        } else {
            self.next_sequential_read_offset = next_request.end_offset();
        }
    }
    if self.current_task.is_none() {
        // No inflight task containing the target offset.
        trace!(current_offset=?self.next_sequential_read_offset, requested_offset=?offset, "seek failed: not enough inflight data");
        return Ok(false);
    }
    // We could try harder to preserve the backwards seek buffer if we're near the
    // request boundary, but it's probably not worth the trouble.
    self.backward_seek_window.clear();
}

let current_task = self
    .current_task
    .as_mut()
    .expect("a request existed that covered this seek offset");

// If we have enough bytes already downloaded (`available`) to skip straight to this read, then do 
// it. Otherwise, we're willing to wait for the bytes to download only if they're coming "soon", where 
// soon is defined as up to `max_forward_seek_distance` bytes ahead of the available offset. 
if offset > current_task.available_offset() + self.config.max_forward_seek_distance { 
    return Ok(false); 
}
```

Reply:
I see 2 decisions which affect functionality, making them explicit: 
1. We seek through tasks which are not fully completed (e.g. `current_task` has downloaded 0.5/1GB, `future_tasks[0]` has downloaded 0.2/1GB, seeked position belongs to `future_tasks[0]` => we drop current and jump to the next)
2. Max offset to wait for is based on `available_offset`: `current_task.available_offset() + self.config.max_forward_seek_distance`, which will make us to wait in more cases.

Reply:
Reasoning in absolute offsets also looks clearer to me. I'll update the PR shortly to do that